homeassistant/components/qnap_qsw/binary_sensor.py

```diff
@@ -43,20 +52,71 @@ class QswBinarySensorEntityDescription(
     ),
 )
 
+LACP_PORT_BINARY_SENSOR_TYPES: Final[tuple[QswBinarySensorEntityDescription, ...]] = (
+    QswBinarySensorEntityDescription(
+        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+        entity_registry_enabled_default=False,
+        key=QSD_PORTS_STATUS,
+        qsw_type=QswEntityType.LACP_PORT,
+        name="Link",
+        subkey=QSD_LINK,
+    ),
+)
+
+PORT_BINARY_SENSOR_TYPES: Final[tuple[QswBinarySensorEntityDescription, ...]] = (
+    QswBinarySensorEntityDescription(
+        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+        entity_registry_enabled_default=False,
+        key=QSD_PORTS_STATUS,
+        qsw_type=QswEntityType.PORT,
+        name="Link",
+        subkey=QSD_LINK,
+    ),
+)
+
 
 async def async_setup_entry(
     hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
     """Add QNAP QSW binary sensors from a config_entry."""
     coordinator: QswDataCoordinator = hass.data[DOMAIN][entry.entry_id][QSW_COORD_DATA]
-    async_add_entities(
-        QswBinarySensor(coordinator, description, entry)
-        for description in BINARY_SENSOR_TYPES
+
+    entities: list[QswBinarySensor] = []
+
+    for description in BINARY_SENSOR_TYPES:
         if (
             description.key in coordinator.data
             and description.subkey in coordinator.data[description.key]
-        )
-    )
+        ):
+            entities.append(QswBinarySensor(coordinator, description, entry))
+
+    for description in LACP_PORT_BINARY_SENSOR_TYPES:
+        if (
+            description.key in coordinator.data
+            and QSD_LACP_PORTS in coordinator.data[description.key]
+        ):
+            for port_id, port_values in coordinator.data[description.key][
+                QSD_LACP_PORTS
+            ].items():
+                if description.subkey in port_values:
+                    entities.append(
+                        QswBinarySensor(coordinator, description, entry, port_id)
+                    )
+
+    for description in PORT_BINARY_SENSOR_TYPES:
+        if (
+            description.key in coordinator.data
+            and QSD_PORTS in coordinator.data[description.key]
+        ):
+            for port_id, port_values in coordinator.data[description.key][
+                QSD_PORTS
+            ].items():
+                if description.subkey in port_values:
+                    entities.append(
+                        QswBinarySensor(coordinator, description, entry, port_id)
```

Comment:
Instead of reusing the descriptions, I would create a description per port_id on the fly using [`replace`](https://docs.python.org/3/library/dataclasses.html#dataclasses.replace)

Reply:
that way you don't need to do any hacky things inside `__init__`.

Reply:
Looks like this is still outstanding

Reply:
Implemented in 35b7ce1570ad814601e0af6bca98dc18b8977f0f
Please, let me know if this matches your suggestion 😉